src/utils.js

```diff
@@ -200,32 +220,167 @@ export const getAgeData = (birthMonth, birthYear, age, ageMonths) => {
+ *
+ * @param {number} RESPONSE_TIME_LOW_THRESHOLD - The minimum acceptable response time threshold in MS.
+ * @param {number} RESPONSE_TIME_HIGH_THRESHOLD - The maximum acceptable response time threshold in MS.
+ * @param {number} RESPONSE_SIMILARITY_THRESHOLD - The maximum acceptable response threshold as a number of identical responses
+ * @param {number} ACCURACY_THRESHOLD - The minimum acceptable accuracy threshold.
+ * @param {array} ignoredReliabilityFlags - An array of flags that should be ignored when evaluating reliability.
+ * @returns {function} baseValidityEvaluator - A function that evaluates the reliability of a run.
+ */
+export function CreateEvaluateValidity({
+  RESPONSE_TIME_LOW_THRESHOLD = 400,
+  RESPONSE_TIME_HIGH_THRESHOLD = 10000,
+  RESPONSE_SIMILARITY_THRESHOLD = 5,
+  ACCURACY_THRESHOLD = 0.2,
+  ignoredReliabilityFlags = ['responseTimeTooSlow', 'responsesTooSimilar', 'accuracyTooLow'],
+}) {
+  return function baseEvaluateValidity({ responseTimes, responses, correct }) {
+    const flags = [];
+
+    // verifies if responseTimes lie above or below a threshold
+    if (median(responseTimes) <= RESPONSE_TIME_LOW_THRESHOLD) {
+      flags.push('responseTimeTooFast');
+    }
+
+    if (median(responseTimes) >= RESPONSE_TIME_HIGH_THRESHOLD) {
+      flags.push('responseTimeTooSlow');
+    }
+
+    const isSimilar =
+      responses.length >= RESPONSE_SIMILARITY_THRESHOLD &&
+      responses
+        .slice(responses.length - RESPONSE_SIMILARITY_THRESHOLD)
+        .every((val, i, arr) => val === arr[0]);
+    // Calculate response similarity based on maxIdenticalResponse
+    if (isSimilar) {
+      flags.push('responsesTooSimilar');
+    }
+
+    // Calculate accuracy based on the number of correct responses
+    const numCorrect = correct?.filter((x) => x === 1).length ?? 0;
+    if (numCorrect / correct.length <= ACCURACY_THRESHOLD) {
+      flags.push('accuracyTooLow');
+    }
+
+    const isReliable = flags.filter((x) => !ignoredReliabilityFlags.includes(x)).length === 0;
+
+    return { flags, isReliable };
+  };
+}
+
+/**
+ * Tracks response times and invokes callback function when response times
+ * exceed specified thresholds.
+ *
+ * @class ResponseTimeTracker
+ * @param {Function} evaluateValidity function to be called to generate an decision on the
+ *    reliability of a run
+ * @param {Function} addEngagementFlags function passed through to update the run's
+ *  firekit object with flags tripped
+ * @param {number} minResponsesRequired The minimum number of responses required before
+ *   checking for threshold exceedance.
+ * @property {Array<number>} _responseTimes An array to store the response times.
+ * @property {Array<number>} _responses An array to store the responses
+ * (keypresses or button choices).
+ * @property {Array<number>} _correct An array to store the correctness of the responses.
+ */
+export class ValidityEvaluator {
+  constructor({
+    evaluateValidity = () => {},
```

Comment:
Can the default here just be the default returned by `createEvaluateValidity()`?

```suggestion
    evaluateValidity = createEvaluateValidity(),
```